### PR TITLE
SCS/SDS/SSS.16 implementation based on WG.2 confirmation -- revoked CA is sub-CA

### PR DESCRIPTION
**Pull request by jni2000**
As WinnF WG.2 and ITS approved to use CRL as the initial certificate blacklisting solution before OCSP is available, the test codes developers working on security related test cases came to the following agreements:

1. Deploy a CRL server (a.k.aThe certificate revocation list distribution point – CDP) in the ITS designated VM environment to serve as a FCC certification dedicated CRL server
2. Certificates, sub-CAs to be blacklisted for all the test cases (e.g. SCS.11 and SCS.16 etc.) will be created and per-loaded into the CRL server appropriately. There is NO need for the test harness codes to blacklist a certificate or sub-CA on-demand. During the test, the test codes will directly use the certificate or sub-CA that are already in the CRL.
3. SAS UUT is assumed to pull the CRL periodically as if in normal operation. There is only a coordination need, that is to make sure SAS UUT has the latest CRL information before running the related test cases.

The SCS.16, SDS.16 and SSS.16 test cases are modified to meeting these agreements as below

1.  the revoked CA is sub-CA that will be inserted into CRL
2. the test cases will use CRL for testing the revoked sub-CA requirement.

### CODE REVIEW

**Comment by kate-harrison** [comment]
I suggest we concentrate on #467 (SCS.11, etc.) and then apply the relevant changes to this PR.

File: src/harness/simple_crl_server.py [hunk 1]
```diff
@@ -26,11 +26,6 @@
 
 DEFAULT_CRL_URL = 'http://localhost:9007/ca.crl'
 DEFAULT_CRL_SERVER_PORT = 80
-CERT_TYPE = {
```
Comment (kate-harrison) [inline src/harness/simple_crl_server.py]: What's the motivation for these changes?

File: src/harness/certs/generate_fake_certs.sh [hunk 2]
```diff
@@ -397,6 +397,45 @@ openssl ca -cert sas_ca.cert -keyfile private/sas_ca.key -in sas.csr \
...
+openssl ca -gencrl -keyfile private/sas_ca.key -cert sas_ca.cert \
+     -config ../../../cert/openssl.cnf -crlhours 1 \
+     -out root/sas_ca.crl
+
+echo "\n\n Generate CRL for proxy_ca"
+openssl ca -gencrl -keyfile private/proxy_ca.key -cert proxy_ca.cert \
+     -config ../../../cert/openssl.cnf -crlhours 1 \
+     -out root/proxy_ca.crl
+
+echo "\n\n Generate CRL for cbsd_ca"
```
Comment (kate-harrison) [inline src/harness/certs/generate_fake_certs.sh]: I can't tell for sure, but it looks like you're creating a new CA and not a new sub-CA for this test case. You should be using the same root CA but creating new sub-CAs (CBSD, DP, SAS) each of which has a client cert. The sub-CAs are the ones that are revoked, not the client certs. The client certs are what should be used in the test case.

File: src/harness/testcases/WINNF_FT_S_SCS_testcase.py [hunk 3]
```diff
@@ -257,6 +257,30 @@ def test_WINNF_FT_S_SCS_15(self,config_filename):
     # Check registration response
     self.assertEqual(response['response']['responseCode'],104)
 
+  def generate_SCS_16_default_config(self, filename):
+    """Generates the WinnForum configuration for SCS_16. """
+    # Create the configuration for client cert/key path.
+
+    config = {
+      'clientCert': self.getCertFilename("client.cert"),
```
Comment (kate-harrison) [inline src/harness/testcases/WINNF_FT_S_SCS_testcase.py]: Is this cert derived from a revoked sub-CA?

File: src/harness/simple_crl_server.py (hunk 1 above)
Comment (jni2000) [reply src/harness/simple_crl_server.py]: To avoid   human error, CRL Server is now enhanced    to identify the cerificate type automatically based on issuer type to   determine whether it is intermediate CA or leaf ceritficate during   certificate revocation.

File: src/harness/certs/generate_fake_certs.sh (hunk 2 above)
Comment (jni2000) [reply src/harness/certs/generate_fake_certs.sh]: Agreed.   We have made required changes to create new sub CA     for CBSD, SAS and Domain Proxy by using existing root CA     revoke the  newly created sub CA and   leaf client certificates are     signed by revoked sub CA.

File: src/harness/testcases/WINNF_FT_S_SCS_testcase.py (hunk 3 above)
Comment (jni2000) [reply src/harness/testcases/WINNF_FT_S_SCS_testcase.py]: No. Now   we have created the new client certificate signed by revoked sub CA.

**Comment by jni2000** [comment]
Kate

We have addressed your latest review comments. Also, we picked up all relevant changes in SxS.11. Please review.

Thanks.

File: src/harness/certs/README.md [hunk 4]
```diff
@@ -24,6 +24,13 @@ https://github.com/Wireless-Innovation-Forum/Spectrum-Access-System/tree/master/
                    sas_expired                                               domain_proxy_expired
                                                                              domain_proxy_inapplicable
 
+                  -------------root_ca --------------------------------------
```
Comment (kate-harrison) [inline src/harness/certs/README.md]: Maybe use `root_ca (same as above, separate graph for clarity)` or something similar, to make it clear that there's only one `root_ca`?

File: src/harness/certs/generate_fake_certs.sh [hunk 5]
```diff
@@ -7,8 +7,11 @@ sed -i '/TEST = critical, ASN1:NULL/d' ../../../cert/openssl.cnf
 rm -rf crl/
 mkdir private
 mkdir root
-touch index.txt
-echo -n 'unique_subject = no' >> index.txt.attr
+mkdir crl
+
+# Empty the index.txt to avoid to old certificates index.
+cat /dev/null > index.txt
```
Comment (kate-harrison) [inline src/harness/certs/generate_fake_certs.sh]: Couldn't you just use `rm`?

File: src/harness/certs/README.md [hunk 6]
```diff
@@ -24,6 +24,13 @@ https://github.com/Wireless-Innovation-Forum/Spectrum-Access-System/tree/master/
                    sas_expired                                               domain_proxy_expired
                                                                              domain_proxy_inapplicable
 
+                  -------------root_ca --------------------------------------                
+                 /                   \         				      \ 
+                /                     \         			       \
+       blacklisted_sas_ca         blacklisted_cbsd_ca   		blacklisted_proxy_ca
```
Comment (kate-harrison) [inline src/harness/certs/README.md]: Probably `revoked_sas_ca` is more accurate (similar for the others).

File: src/harness/certs/README.md [hunk 7]
```diff
@@ -24,6 +24,13 @@ https://github.com/Wireless-Innovation-Forum/Spectrum-Access-System/tree/master/
...
+                \                        \                                   \
+                 \                        \                                   \
+            sas_revoked_by_ca         client_revoked_by_ca              domain_proxy_revoked_by_ca
```
Comment (kate-harrison) [inline src/harness/certs/README.md]: I think `sas_cert_from_revoked_ca` is more accurate. The cert itself isn't revoked, but its issuer is.

File: src/harness/certs/README.md (hunk 4 above)
Comment (jni2000) [reply src/harness/certs/README.md]: Update   the graph based on the suggestion

File: src/harness/certs/generate_fake_certs.sh (hunk 5 above)
Comment (jni2000) [reply src/harness/certs/generate_fake_certs.sh]: sub CA's   are named correctly to give better clarity

File: src/harness/certs/README.md (hunk 6 above)
Comment (jni2000) [reply src/harness/certs/README.md]: sub CA's   are named correctly to give better clarity

File: src/harness/certs/README.md (hunk 7 above)
Comment (jni2000) [reply src/harness/certs/README.md]: leaf   certificate names are updated correctly to give clear indication signed by   revoked CA.

**Comment by jni2000** [comment]
Kate, we have addressed your review comments. Please continue reviewing. Thanks.

File: src/harness/certs/README.md [hunk 8]
```diff
@@ -24,6 +24,13 @@ https://github.com/Wireless-Innovation-Forum/Spectrum-Access-System/tree/master/
                    sas_expired                                               domain_proxy_expired
                                                                              domain_proxy_inapplicable
 
+                  -------------root_ca (same as above, separate graph for the clarity)--------------------
+                 /                    \         		     \
```
Comment (pmarks-net) [inline src/harness/certs/README.md]: Please remove the tab characters from this diagram.

File: src/harness/certs/revoke_and_generate_crl.sh [hunk 9]
```diff
@@ -2,53 +2,79 @@
...
 {
-  #Argument1: Name of the certificate to be  blacklisted.
-  #Argument2: Intermediate CA to generate CRL.
+  #Argument1: Name of the certificate to be revoked.
 
+  #Fetch the Common Name to find out the issuer
+  CN=`openssl x509 -issuer -in $1 -noout | sed 's/^.*CN=//'`
+
+  if [ "$CN" == "WInnForum RSA CBSD CA-1" ]; then
+      CA=cbsd_ca
```
Comment (pmarks-net) [inline src/harness/certs/revoke_and_generate_crl.sh]: This file uses a 2-space indent for nested blocks.  Please maintain that.

File: src/harness/certs/revoke_and_generate_crl.sh [hunk 10]
```diff
@@ -2,53 +2,79 @@
...
+  elif [ "$CN" == "WInnForum CBSD CA-1 - Revoked" ]; then
+      CA=revoked_cbsd_ca
+  elif [ "$CN" == "WInnForum RSA Domain Proxy CA - Revoked" ]; then
+      CA=revoked_proxy_ca
+  elif [ "$CN" == "WInnForum RSA SAS CA-1 - Revoked" ]; then
+      CA=revoked_sas_ca
+  else
+      echo "Unknown issuer CN=$CN for certificate $1"
+      exit -1
+   fi
```
Comment (pmarks-net) [inline src/harness/certs/revoke_and_generate_crl.sh]: Indent -1 space.

File: src/harness/certs/revoke_and_generate_crl.sh [hunk 11]
```diff
@@ -2,53 +2,79 @@
...
 
-function blacklist_certificate()
+function revoke_certificate()
 {
-  #Argument1: Name of the certificate to be  blacklisted.
-  #Argument2: Intermediate CA to generate CRL.
+  #Argument1: Name of the certificate to be revoked.
 
+  #Fetch the Common Name to find out the issuer
+  CN=`openssl x509 -issuer -in $1 -noout | sed 's/^.*CN=//'`
```
Comment (pmarks-net) [inline src/harness/certs/revoke_and_generate_crl.sh]: `CN` and `CA` should be [local variables](http://tldp.org/LDP/abs/html/localvar.html).

File: src/harness/simple_crl_server.py [hunk 12]
```diff
@@ -182,17 +177,8 @@ def revokeCertificate():
...
-                  for (cert_id, cert_type) in sorted(CERT_TYPE.iteritems()))
-  cert_type = raw_input(CLI_PROMPT_TEXT)
-  if cert_type not in CERT_TYPE:
-    raise Exception("RunTimeError:Type of certificate to be blacklisted is incorrect."
-                    "Please select one of the values:%s" % CERT_TYPE.values())
-
-  revoke_cert_command = "cd {0} && ./revoke_and_generate_crl.sh " \
-                        "{1} {2}".format(getCertsDirectoryAbsolutePath(),
-                                         CERT_TYPE[cert_type],
+  revoke_cert_command = "cd {0} && bash ./revoke_and_generate_crl.sh " \
```
Comment (pmarks-net) [inline src/harness/simple_crl_server.py]: What's the motivation behind the explicit `bash` here?  The script already starts with `#!/bin/bash`.

File: src/harness/certs/README.md (hunk 8 above)
Comment (jni2000) [reply src/harness/certs/README.md]: Typo has   been fixed

File: src/harness/certs/revoke_and_generate_crl.sh (hunk 9 above)
Comment (jni2000) [reply src/harness/certs/revoke_and_generate_crl.sh]: Fixed the indent issue

File: src/harness/certs/revoke_and_generate_crl.sh (hunk 10 above)
Comment (jni2000) [reply src/harness/certs/revoke_and_generate_crl.sh]: Fixed

File: src/harness/certs/revoke_and_generate_crl.sh (hunk 11 above)
Comment (jni2000) [reply src/harness/certs/revoke_and_generate_crl.sh]: Changed   as local variables

File: src/harness/simple_crl_server.py (hunk 12 above)
Comment (jni2000) [reply src/harness/simple_crl_server.py]: Yes   revert it back to previous state

**Comment by jni2000** [comment]
Paul

We have addressed all your latest review comments, please take a look. Thanks.
